os/board/rtl8720e/src/component/bluetooth/api/ipc/bt_ipc_parameter_conversion.c

```diff
@@ -25,17 +31,126 @@
+{
+	if (act == RTK_BT_DEVICE_IPC_ACT_BT_ENABLE) {
+#if (defined RTK_BLE_AUDIO_SUPPORT) && (RTK_BLE_AUDIO_SUPPORT)	
+		rtk_bt_app_conf_t *param = (rtk_bt_app_conf_t *)dst;
+		rtk_bt_le_audio_app_conf_t *le_audio_app_conf = &param->le_audio_app_conf;
+		*le_audio_app_conf->pacs_param.p_sink_pac_id = *((int8_t *)src);
+		*le_audio_app_conf->pacs_param.p_pac_sink_codec = *((int8_t *)src + 1);
+		*le_audio_app_conf->pacs_param.p_source_pac_id = *((int8_t *)src + 2);
+		*le_audio_app_conf->pacs_param.p_pac_source_codec = *((int8_t *)src + 3);
+		if(le_audio_app_conf->profile_support & RTK_BT_LE_AUDIO_PROFILE_MCS)
+			*le_audio_app_conf->mcs_param.p_mcs_id = *((int8_t *)src + 4);
+#else 
+		(void)src;
+		(void)dst;
+#endif
+	}
+	return RTK_BT_OK;
+}
+
+/* ===============================================rtk_bt_gap ipc action======================================= */
+uint16_t rtk_gap_copy_act_ret_param(uint16_t act, void *dst, void *src)
+{
+	(void)dst;
+	(void)src;
+
+	switch(act) {
+#if RTK_BT_5_2_L2C_ECFC_SUPPORT
+	case RTK_BT_GAP_ACT_ECFC_REGISTER: {
+		rtk_bt_gap_ecfc_register_t *param = (rtk_bt_gap_ecfc_register_t *)dst;
+		memcpy(param->p_proto_id, (void *)src, sizeof(uint16_t));
+		break;
+	}
+#endif
+
+	default:
+		break;
+	}
+
+	return RTK_BT_OK;
+}
+
+#if RTK_BT_5_2_L2C_ECFC_SUPPORT
+static int gap_ipc_ecfc_send_data_push(uint8_t *buf, rtk_bt_gap_ecfc_send_data_t *param, 
+									   uint32_t size_limit, uint32_t *actual_size)
+{
+	uint8_t *end = buf;
+	uint8_t *struct_end = NULL;
+	uint32_t size = 0;
+
+	/* memory space for rtk_bt_gatts_ntf_and_ind_param_t */
+	size = sizeof(rtk_bt_gap_ecfc_send_data_t);
+	if (end + size > buf + size_limit)
+		goto exit;
+	memcpy(end, param, size);
+	end += ALIGN_UP(size, 4);
+	struct_end = end;
+
+	size = param->length;
+	if (end + size > buf + size_limit)
+		goto exit;
+	memcpy(end, param->p_data, size);
+	end += ALIGN_UP(size, 4);
+
+	((rtk_bt_gap_ecfc_send_data_t *)buf)->p_data = (uint8_t *)(struct_end - buf);
+	*actual_size = (uint32_t)(end - buf);
+	return 0;
+
+exit:
+	printf("gap_ipc_ecfc_send_data_filter_push size is over deep copy buf\r\n");
+	return -1;
+}
+#endif
+
+void *bt_gap_push_cmd_ipc_buf(uint16_t act, void *data, uint32_t buf_len, uint32_t *pactual_size)
+{
+	int ret = 0;
+	bt_ipc_host_request_message *host_msg = osif_mem_alloc(RAM_TYPE_DATA_ON, sizeof(bt_ipc_host_request_message));
+
+	if (!host_msg) {
+		printf("%s allocate hsot_msg fail \r\n", __func__);
+		return NULL;
+	}
+	switch (act) {
+#if RTK_BT_5_2_L2C_ECFC_SUPPORT
+	case RTK_BT_GAP_ACT_ECFC_SEND_DATA:
+		ret = gap_ipc_ecfc_send_data_push((uint8_t *)host_msg->param_buf, (rtk_bt_gap_ecfc_send_data_t *)data, 
+									IPC_HOST_API_DATA_MAX, pactual_size);
+		break;
+#endif
+	default:
+		memcpy((void *)host_msg->param_buf, (void *)data, buf_len);
+		*pactual_size = buf_len;
+		break;
+	}
+
+	if (ret) {
+		osif_mem_free(host_msg);
+		host_msg = NULL;
+	}
+
+	return (void *)host_msg;

```

Comment:
Could you double-confirm this will be freed at caller?

Reply:
@sunghan-chang yes this will be freed at line 112 of rtk_bt_gap_set_act function, and the subsequent functions.
![image](https://user-images.githubusercontent.com/19281551/217990236-af34d862-be75-4c7a-bfd7-b077b1e6347e.png)